bundle/config/loader/process_include.go

```diff
@@ -3,12 +3,144 @@ package loader
+	"cluster",
+}
+
+func validateFileFormat(r *config.Root, filePath string) diag.Diagnostics {
+	for _, typ := range resourceTypes {
+		for _, ext := range []string{fmt.Sprintf(".%s.yml", typ), fmt.Sprintf(".%s.yaml", typ)} {
+			if strings.HasSuffix(filePath, ext) {
+				return validateSingleResourceDefined(r, ext, typ)
+			}
+		}
+	}
+
+	return nil
+}
+
+func validateSingleResourceDefined(r *config.Root, ext, typ string) diag.Diagnostics {
+	type resource struct {
+		path  dyn.Path
+		value dyn.Value
+		typ   string
+		key   string
+	}
+
+	resources := []resource{}
+
+	// Gather all resources defined in the resources block.
+	_, err := dyn.MapByPattern(
+		r.Value(),
+		dyn.NewPattern(dyn.Key("resources"), dyn.AnyKey(), dyn.AnyKey()),
+		func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			// The key for the resource. Eg: "my_job" for jobs.my_job.
+			k := p[2].Key()
+			// The type of the resource. Eg: "job" for jobs.my_job.
+			typ := strings.TrimSuffix(p[1].Key(), "s")
+
+			resources = append(resources, resource{path: p, value: v, typ: typ, key: k})
+			return v, nil
+		})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// Gather all resources defined in a target block.
+	_, err = dyn.MapByPattern(
+		r.Value(),
+		dyn.NewPattern(dyn.Key("targets"), dyn.AnyKey(), dyn.Key("resources"), dyn.AnyKey(), dyn.AnyKey()),
+		func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			// The key for the resource. Eg: "my_job" for jobs.my_job.
+			k := p[4].Key()
+			// The type of the resource. Eg: "job" for jobs.my_job.
+			typ := strings.TrimSuffix(p[3].Key(), "s")
+
+			resources = append(resources, resource{path: p, value: v, typ: typ, key: k})
+			return v, nil
+		})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	typeMatch := true
+	seenKeys := map[string]struct{}{}
+	for _, rr := range resources {
+		// case: The resource is not of the correct type.
+		if rr.typ != typ {
+			typeMatch = false
+			break
+		}
+
+		seenKeys[rr.key] = struct{}{}
+	}
+
+	// Format matches. There's less than or equal to one resource defined in the file.
+	// The resource is also of the correct type.
+	if typeMatch && len(seenKeys) <= 1 {
+		return nil
+	}
+
+	msg := strings.Builder{}
+	msg.WriteString(fmt.Sprintf("We recommend only defining a single %s in a file with the %s extension.\n", typ, ext))
+
+	// Dedup the list of resources before adding them the diagnostic message. This
+	// is needed because we do not dedup earlier when gathering the resources and
+	// it's valid to define the same resource in both the resources and targets block.
+	msg.WriteString("The following resources are defined or configured in this file:\n")
+	setOfLines := map[string]struct{}{}
+	for _, r := range resources {
+		setOfLines[fmt.Sprintf("  - %s (%s)\n", r.key, r.typ)] = struct{}{}
+	}
+	// Sort the line s to print to make the output deterministic.
+	listOfLines := maps.Keys(setOfLines)
+	sort.Strings(listOfLines)
+	for _, l := range listOfLines {
+		msg.WriteString(l)
+	}
```

Comment:
You can use a string slice directly, sort it, and then call `slices.Compact` to remove dups.

It's a bit odd to use a map here only to dedup entries.

Reply:
done